packages/SwingSet/misc-tools/db-clone.js

```diff
@@ -0,0 +1,529 @@
+
+  try {
+    srcDb.prepare('BEGIN TRANSACTION').run();
+    destDb.prepare('BEGIN TRANSACTION').run();
+
+    if (stats) {
+      getStats(srcDb);
+      getStats(destDb);
+    }
+
+    // filter transactions first so bundle/snapshot filters can be populated for "auto" option
+    if (isSet(transcripts)) {
+      didWork = true;
+      log('Copying transcripts...');
+      if (transcripts === 'all' && (startFilter || endFilter))
+        warn('--transcripts=all overrides --startPos / --endPos');
+
+      let items = [];
+      let spans = [];
+
+      // skip query if we have an empty set of vats
+      if (vatFilter === null || vatFilter.size !== 0) {
+        const sqlVatFilter = vatFilter
+          ? `AND vatID IN (${[...vatFilter].map(() => '?').join(',')})`
+          : '';
+
+        const transcriptItemsQuery = `SELECT * FROM transcriptItems WHERE 1=1
+            ${sqlVatFilter}
+            ${startFilter ? 'AND position >= ?' : ''}
+            ${endFilter ? 'AND position <= ?' : ''}`;
+
+        const transcriptSpansQuery = `SELECT * FROM transcriptSpans WHERE 1=1
+            ${sqlVatFilter}
+            ${startFilter ? 'AND endPos >= ?' : ''}
+            ${endFilter ? 'AND startPos <= ?' : ''}`;
+
+        dbg(`transcripts = ${transcripts};`);
+        dbg(`vatFilter.size = ${vatFilter?.size};`);
+        dbg(`startFilter = ${startFilter};`);
+        dbg(`endFilter = ${endFilter};`);
+        dbg(`transcriptItemsQuery = ${transcriptItemsQuery};`);
+        dbg(`transcriptSpansQuery = ${transcriptSpansQuery};`);
+
+        const sqlArgs = [
+          ...(vatFilter || []),
+          ...(startFilter ? [startFilter] : []),
+          ...(endFilter ? [endFilter] : []),
+        ].filter(x => isSet(x)); // allow for zeroes
+
+        items = srcDb.prepare(transcriptItemsQuery).all(...sqlArgs);
+        spans = srcDb.prepare(transcriptSpansQuery).all(...sqlArgs);
+      }
+
+      // create tables regardless of whether any items had been found or not
+      destDb.exec(`CREATE TABLE IF NOT EXISTS transcriptItems (
+        vatID TEXT, position INTEGER, item TEXT, incarnation INTEGER, PRIMARY KEY (vatID, position)
+      )`);
+      destDb.exec(`CREATE TABLE IF NOT EXISTS transcriptSpans (
+        vatID TEXT, startPos INTEGER, endPos INTEGER, hash TEXT, isCurrent INTEGER CHECK (isCurrent = 1), incarnation INTEGER,
+        PRIMARY KEY (vatID, startPos), UNIQUE (vatID, isCurrent)
+      )`);
+
+      destDb.exec(
+        `CREATE INDEX IF NOT EXISTS currentTranscriptIndex ON transcriptSpans (vatID, isCurrent)`,
+      );
+
+      const insertTranscriptItems = destDb.prepare(
+        `INSERT OR IGNORE INTO transcriptItems (vatID, position, item, incarnation) VALUES (?, ?, ?, ?)`,
+      );
+      const insertTranscriptSpans = destDb.prepare(
+        `INSERT OR IGNORE INTO transcriptSpans (vatID, startPos, endPos, hash, isCurrent, incarnation) VALUES (?, ?, ?, ?, ?, ?)`,
+      );
+
```

Comment:
I'm worried about maintenance. Is there any way to replicate the schema of the source db ?

Reply:
Yes and no.
Yes: well, it is obviously available in the source DB itself.
No: we are filtering and looking at very particular columns to massage the data, which implies that we do know the schema upfront.

and now back to yes: this low-level data massaging should be abstracted away into a separate layer, probably below the *Store layer. Ideally it should also allow for SQL-native copy between the databases without having to go through JavaSctipt. See [ATTACH DATABASE](https://www.sqlitetutorial.net/sqlite-attach-database/) for example.